README
gst-bs2b
========

This uses the bs2b library, created by Boris Mikhaylov [1] and is based on
the GStreamer plugin template [2].

Settings:

  fcut: Low-pass filter cut frequency (Hz) - 300..2000
  feed: Feed Level (dB/10) - 10..150 (1..15 dB)
  active: Specify whether the filter is active - True/False

Presets:

  (0): default
    Closest to virtual speaker placement (30°, 3 meter)   [700Hz, 4.5dB]
  (1): cmoy
    Close to Chu Moy's crossfeeder (popular)              [700Hz, 6.0dB]
  (2): jmeier
    Close to Jan Meier's CORDA amplifiers (little change) [650Hz, 9.0dB]
  (3): none
    No preset

Example pipeline:

  gst-launch -v audiotestsrc ! interleave name=i ! crossfeed ! \
    autoaudiosink audiotestsrc freq=330 ! i.

[1] http://bs2b.sourceforge.net/
[2] http://cgit.freedesktop.org/gstreamer/gst-template/

License
=======

Most of the autotool files are MIT [1] licensed and the plugin code
is LGPLv2 (or later) [2] licensed.
See the files for exact licensing information.

[1] COPYING.MIT
[2] COPYING

Building
========

./autogen.sh
make

You can also check if it has been built correctly with:

  gst-inspect-0.10 src/.libs/libgstbs2b.so

./autogen-clean.sh

  to removes all autogenerated files

./test.py (needs GStreamer python bindings)

  test the element


NEWS
0.4 (Mon, 29 Aug 2011 19:22:52 +0200)
    * Fix leak
    * Property descriptions, preset -> enum, feed is now dB/10 not dB
    * Clear filter buffer on new segments (after seek, reactivating etc.)
    * 24 bit and single channel support
    * Various fixes

0.3 (Sat, 20 Aug 2011 03:19:57 +0200)

    * First real release


test.py
#!/usr/bin/env python
# Copyright 2011 Christoph Reiter <christoph.reiter@gmx.at>
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Library General Public
# License as published by the Free Software Foundation; either
# version 2 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Library General Public License for more details.
#
# You should have received a copy of the GNU Library General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA

import os
import sys

env_key = "GST_PLUGIN_PATH"
build_path = os.path.join(
    os.path.dirname(os.path.realpath(sys.argv[0])), "src", ".libs")
if not env_key in os.environ:
    print "Info: '%s' not set, using default: '%s'" % (env_key, build_path)
    so_path = os.path.join(build_path, "libgstbs2b.so")
    if not os.path.exists(so_path):
        print "Warning: Could not find '%s'. Did you type 'make'?" % so_path
        sys.exit(1)
    os.environ[env_key] = build_path

import unittest
import threading

import gst
import gobject

gobject.threads_init()

# These are missing in gobject
gobject.PARAM_STATIC_NAME = 1 << 5
gobject.PARAM_STATIC_NICK = 1 << 6
gobject.PARAM_STATIC_BLURB = 1 << 7
gobject.PARAM_STATIC_STRINGS = (
    gobject.PARAM_STATIC_NAME |
    gobject.PARAM_STATIC_NICK |
    gobject.PARAM_STATIC_BLURB)

class CrossfeedProperties(unittest.TestCase):
    def setUp(self):
        self.crossfeed = gst.element_factory_make('crossfeed')
        self.props = gobject.list_properties(self.crossfeed)
        self.props = dict(((p.name, p) for p in self.props))

    def test_names_nicks(self):
        for name in ["fcut", "feed", "preset", "active"]:
            self.assertTrue(name in self.props)

        nicks = [p.nick for p in self.props.values()]
        for name in ["Frequency cut", "Feed level", "Preset", "Active"]:
            self.assertTrue(name in nicks)

        for i, nick in enumerate(["default", "cmoy", "jmeier"]):
            self.crossfeed.set_property("preset", i)
            self.assertEqual(
                self.crossfeed.get_property("preset").value_nick, nick)

        for i, name in enumerate(["speaker", "Chu Moy", "Jan Meier"]):
            self.crossfeed.set_property("preset", i)
            self.assertTrue(name in
                self.crossfeed.get_property("preset").value_name)

        self.crossfeed.set_property("fcut", 999)
        self.assertEqual(
            self.crossfeed.get_property("preset").value_nick, "none")
        self.assertTrue("No " in
            self.crossfeed.get_property("preset").value_name)

    def test_default(self):
        self.assertEqual(self.props["active"].default_value, True)
        self.assertEqual(self.props["fcut"].default_value, 700)
        self.assertEqual(self.props["feed"].default_value, 45)
        self.assertEqual(self.props["preset"].default_value, 0)
        self.assertEqual(
            self.props["preset"].default_value.value_nick, "default")

    def test_real_default(self):
        self.assertEqual(self.crossfeed.get_property("active"), True)
        self.assertEqual(self.crossfeed.get_property("fcut"), 700)
        self.assertEqual(self.crossfeed.get_property("feed"), 45)
        self.assertEqual(self.crossfeed.get_property("preset").real, 0)

    def test_range(self):
        self.assertEqual(self.props["fcut"].maximum, 2000)
        self.assertEqual(self.props["fcut"].minimum, 300)

        self.assertEqual(self.props["feed"].maximum, 150)
        self.assertEqual(self.props["feed"].minimum, 10)

    def test_types(self):
        self.assertEqual(self.props["active"].value_type, gobject.TYPE_BOOLEAN)
        self.assertEqual(self.props["fcut"].value_type, gobject.TYPE_INT)
        self.assertEqual(self.props["feed"].value_type, gobject.TYPE_INT)
        self.assertEqual(
            self.props["preset"].value_type.parent, gobject.TYPE_ENUM)

    def test_flags(self):
        for name in ["active", "feed", "fcut", "preset"]:
            self.assertEqual(self.props[name].flags,
                gobject.PARAM_READWRITE |
                gobject.PARAM_STATIC_STRINGS |
                gst.PARAM_CONTROLLABLE)

    def dearDown(self):
        del self.crossfeed

class CrossfeedCaps(unittest.TestCase):
    def setUp(self):
        self.mainloop = gobject.MainLoop()
        self.p = None

    def __stop(self, error):
        if self.p:
            self.p.set_state(gst.STATE_NULL)
            bus = self.p.get_bus()
            bus.disconnect(self.bus_id)
            bus.remove_signal_watch()
            self.p = None
        self.error = error
        self.mainloop.quit()

    def __start(self, t):
        self.error = False
        try: self.p = gst.parse_launch("audiotestsrc num-buffers=42 ! " + t)
        except gobject.GError:
            self.__stop(True)
            return

        bus = self.p.get_bus()
        bus.add_signal_watch()
        bus.enable_sync_message_emission()
        self.bus_id = bus.connect("sync-message", self.__message)

        self.p.set_state(gst.STATE_PLAYING)

    def __message(self, bus, message):
        if message.type == gst.MESSAGE_EOS:
            gobject.idle_add(self.__stop, False)
        elif message.type == gst.MESSAGE_ERROR:
            gobject.idle_add(self.__stop, True)

    def __check(self, t):
        gobject.idle_add(self.__start, t)
        self.mainloop.run()
        return not self.error

    def test_basic(self):
        r = self.__check("crossfeed ! fakesink")
        self.assertTrue(r)

        r = self.__check(
            "audioconvert ! audio/x-raw-int,width=8,depth=8 ! "
            "crossfeed ! fakesink")
        self.assertTrue(r)

    def test_formats(self):
        for width in [8, 16, 24, 32]:
            r = self.__check(
                "audioconvert ! audio/x-raw-int,width=%d,depth=%d ! "
                "crossfeed ! fakesink" % (width, width))
            self.assertTrue(r)

        for width in [32, 64]:
            r = self.__check(
                "audioconvert ! audio/x-raw-float,width=%d,depth=%d ! "
                "crossfeed ! fakesink" % (width, width))
            self.assertTrue(r)

        for depth in range(1, 65, 7):
            r = self.__check(
                "audioconvert ! audio/x-raw-float,width=%d,depth=%d ! "
                "crossfeed ! fakesink" % (64, depth))
            self.assertTrue(r)

        for depth in [8, 16, 24, 32]:
            r = self.__check(
                "audioconvert ! audio/x-raw-int,width=%d,depth=%d ! "
                "crossfeed ! fakesink" % (32, depth))
            self.assertTrue(r)

    def test_channels(self):
        r = self.__check(
            "audioconvert ! audio/x-raw-int,channels=1 ! "
            "crossfeed ! fakesink")
        self.assertTrue(r)

        r = self.__check(
            "audioconvert ! audio/x-raw-int,channels=2 ! "
            "crossfeed ! fakesink")
        self.assertTrue(r)

        r = self.__check(
            "audioconvert ! audio/x-raw-int,channels=3 ! "
            "crossfeed ! fakesink")
        self.assertFalse(r)

    def test_endianness(self):
        r = self.__check(
            "audioconvert ! audio/x-raw-int,endianess=1234 ! "
            "crossfeed ! fakesink")
        self.assertTrue(r)

        r = self.__check(
            "audioconvert ! audio/x-raw-int,endianess=4321 ! "
            "crossfeed ! fakesink")
        self.assertTrue(r)

    def test_rate(self):
        r = self.__check(
            "audioconvert ! audio/x-raw-int,rate=2000 ! "
            "crossfeed ! fakesink")
        self.assertTrue(r)

        r = self.__check(
            "audioconvert ! audio/x-raw-int,rate=384000 ! "
            "crossfeed ! fakesink")
        self.assertTrue(r)

        r = self.__check(
            "audioconvert ! audio/x-raw-int,rate=1999 ! "
            "crossfeed ! fakesink")
        self.assertFalse(r)

        r = self.__check(
            "audioconvert ! audio/x-raw-int,rate=384001 ! "
            "crossfeed ! fakesink")
        self.assertFalse(r)

if __name__ == "__main__":
    unittest.main()


autogen.sh
#!/bin/sh
#
# Permission is hereby granted, free of charge, to any person obtaining a
# copy of this software and associated documentation files (the "Software"),
# to deal in the Software without restriction, including without limitation
# the rights to use, copy, modify, merge, publish, distribute, sublicense,
# and/or sell copies of the Software, and to permit persons to whom the
# Software is furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
# DEALINGS IN THE SOFTWARE.

autoreconf --verbose --force --install --make || {
 echo 'autogen.sh failed';
 exit 1;
}

./configure || {
 echo 'configure failed';
 exit 1;
}

echo
echo "Now type 'make' to compile this module."
echo


src/gstbs2b.c
/* GStreamer
 * Copyright (C) <1999> Erik Walthinsen <omega@cse.ogi.edu>
 * Copyright (C) <2003> David Schleef <ds@schleef.org>
 * Copyright (C) <2011> Christoph Reiter <christoph.reiter@gmx.at>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 */

/**
 * SECTION:element-crossfeed
 *
 * Improve headphone listening of stereo audio records using the bs2b library.
 *
 * <refsect2>
 * <title>Example pipelines</title>
 * |[
 * gst-launch -v audiotestsrc ! interleave name=i ! crossfeed ! autoaudiosink audiotestsrc freq=330 ! i.
 * ]| Play two independent sine test sources and crossfeed them.
 * </refsect2>
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <gst/gst.h>
#include <gst/audio/audio.h>
#include <gst/audio/gstaudiofilter.h>

#include <bs2b/bs2b.h>

typedef struct _GstCrossfeed GstCrossfeed;
typedef struct _GstCrossfeedClass GstCrossfeedClass;

#define GST_TYPE_CROSSFEED \
  (gst_crossfeed_get_type())
#define GST_CROSSFEED(obj) \
  (G_TYPE_CHECK_INSTANCE_CAST((obj),GST_TYPE_CROSSFEED,GstCrossfeed))
#define GST_CROSSFEED_CLASS(klass) \
  (G_TYPE_CHECK_CLASS_CAST((klass),GST_TYPE_CROSSFEED,GstCrossfeedClass))
#define GST_IS_CROSSFEED(obj) \
  (G_TYPE_CHECK_INSTANCE_TYPE((obj),GST_TYPE_CROSSFEED))
#define GST_IS_CROSSFEED_CLASS(klass) \
  (G_TYPE_CHECK_CLASS_TYPE((klass),GST_TYPE_CROSSFEED))

#define GST_CROSSFEED_BS2B_LOCK(obj) g_mutex_lock (obj->bs2b_lock)
#define GST_CROSSFEED_BS2B_UNLOCK(obj) g_mutex_unlock (obj->bs2b_lock)

#define GST_CROSSFEED_LOCK(obj) g_mutex_lock (obj->lock)
#define GST_CROSSFEED_UNLOCK(obj) g_mutex_unlock (obj->lock)

struct _GstCrossfeed {
  GstAudioFilter element;

  GMutex *bs2b_lock;
  t_bs2bdp bs2bdp;
  void (*func) ();

  GMutex *lock;
  gboolean active;
  gint divider;
};

struct _GstCrossfeedClass {
  GstAudioFilterClass parent_class;
};

static const GstElementDetails crossfeed_details = GST_ELEMENT_DETAILS (
    "Crossfeed effect",
    "Filter/Effect/Audio",
    "Improve headphone listening of stereo audio records"
        "using the bs2b library.",
    "Christoph Reiter <christoph.reiter@gmx.at>");

#define ALLOWED_CAPS \
    "audio/x-raw-int, " \
    "rate = (int) [ " \
        G_STRINGIFY (BS2B_MINSRATE) "," G_STRINGIFY (BS2B_MAXSRATE) \
    " ], " \
    "channels = (int) { 1, 2 }, " \
    "endianness = (int) { 1234, 4321 }, " \
    "width = (int) { 8, 16, 24, 32 }, " \
    "signed = (boolean) { true, false }; " \
    "audio/x-raw-float, " \
    "rate = (int) [ " \
        G_STRINGIFY (BS2B_MINSRATE) "," G_STRINGIFY (BS2B_MAXSRATE) \
    " ], " \
    "channels = (int) { 1, 2 }, " \
    "endianness = (int) { 1234, 4321 }, " \
    "width = (int) {32, 64} "

enum
{
  ARG_0,
  ARG_ACTIVE,
  ARG_FCUT,
  ARG_FEED,
  ARG_PRESET
};

typedef enum
{
  PRESET_DEFAULT,
  PRESET_CMOY,
  PRESET_JMEIER,
  PRESET_NONE
} GstBs2bPreset;

#define DEFAULT_FCUT (BS2B_DEFAULT_CLEVEL & 0xFFFF)
#define DEFAULT_FEED (BS2B_DEFAULT_CLEVEL >> 16)

static void gst_crossfeed_set_property (GObject * object, guint prop_id,
    const GValue * value, GParamSpec * pspec);
static void gst_crossfeed_get_property (GObject * object, guint prop_id,
    GValue * value, GParamSpec * pspec);
static GType gst_crossfeed_preset_get_type (void);

static gboolean gst_crossfeed_setup (GstAudioFilter * self,
    GstRingBufferSpec * format);

static void gst_crossfeed_finalize (GObject * object);

static gboolean gst_crossfeed_sink_eventfunc (GstBaseTransform * trans,
    GstEvent * event);

static GstFlowReturn gst_crossfeed_transform_inplace (GstBaseTransform * base,
    GstBuffer * outbuf);

static void gst_crossfeed_update_passthrough (GstCrossfeed * crossfeed);

GST_BOILERPLATE (GstCrossfeed, gst_crossfeed, GstAudioFilter,
    GST_TYPE_AUDIO_FILTER);

static void
gst_crossfeed_base_init (gpointer g_class)
{
  GstElementClass *element_class = GST_ELEMENT_CLASS (g_class);
  GstCaps *caps;

  gst_element_class_set_details (element_class, &crossfeed_details);

  caps = gst_caps_from_string (ALLOWED_CAPS);
  gst_audio_filter_class_add_pad_templates (GST_AUDIO_FILTER_CLASS (g_class),
      caps);
  gst_caps_unref (caps);
}

static void
gst_crossfeed_class_init (GstCrossfeedClass * klass)
{
  GObjectClass *gobject_class = (GObjectClass *) klass;
  GstBaseTransformClass *trans_class = (GstBaseTransformClass *) klass;
  GstAudioFilterClass *filter_class = (GstAudioFilterClass *) klass;

  gobject_class->set_property = gst_crossfeed_set_property;
  gobject_class->get_property = gst_crossfeed_get_property;
  gobject_class->finalize = gst_crossfeed_finalize;

  trans_class->transform_ip = gst_crossfeed_transform_inplace;
  trans_class->event = gst_crossfeed_sink_eventfunc;
  filter_class->setup = gst_crossfeed_setup;

  g_object_class_install_property (gobject_class, ARG_ACTIVE,
      g_param_spec_boolean ("active", "Active",
          "Specify whether the filter is active",
          TRUE, G_PARAM_READWRITE | GST_PARAM_CONTROLLABLE |
          G_PARAM_STATIC_STRINGS));

  g_object_class_install_property (gobject_class, ARG_FCUT,
      g_param_spec_int ("fcut", "Frequency cut",
          "Low-pass filter cut frequency (Hz)",
          BS2B_MINFCUT, BS2B_MAXFCUT, DEFAULT_FCUT,
          G_PARAM_READWRITE | GST_PARAM_CONTROLLABLE |
          G_PARAM_STATIC_STRINGS));

  g_object_class_install_property (gobject_class, ARG_FEED,
      g_param_spec_int ("feed", "Feed level", "Feed Level (dB/10)",
          BS2B_MINFEED, BS2B_MAXFEED, DEFAULT_FEED,
          G_PARAM_READWRITE | GST_PARAM_CONTROLLABLE |
          G_PARAM_STATIC_STRINGS));

  g_object_class_install_property (gobject_class, ARG_PRESET,
      g_param_spec_enum ("preset", "Preset", "Bs2b filter preset",
          gst_crossfeed_preset_get_type (),
          PRESET_DEFAULT, G_PARAM_READWRITE | GST_PARAM_CONTROLLABLE |
          G_PARAM_STATIC_STRINGS));
}

static void
gst_crossfeed_init (GstCrossfeed * crossfeed, GstCrossfeedClass * klass)
{
  gst_base_transform_set_gap_aware (GST_BASE_TRANSFORM (crossfeed), TRUE);

  crossfeed->lock = g_mutex_new();
  crossfeed->bs2b_lock = g_mutex_new();
  crossfeed->active = TRUE;

  GST_CROSSFEED_BS2B_LOCK (crossfeed);
  crossfeed->bs2bdp = bs2b_open ();
  GST_CROSSFEED_BS2B_UNLOCK (crossfeed);
}

static gboolean gst_crossfeed_setup (GstAudioFilter * filter,
  GstRingBufferSpec * format)
{
  GstCrossfeed *crossfeed = GST_CROSSFEED (filter);

  crossfeed->func = NULL;

  if (format->type == GST_BUFTYPE_LINEAR) {
    if (format->width == 8) {
      if (format->sign)
        crossfeed->func = &bs2b_cross_feed_s8;
      else
        crossfeed->func = &bs2b_cross_feed_u8;
    }
    else if (format->width == 16) {
      if (format->bigend && format->sign)
        crossfeed->func = &bs2b_cross_feed_s16be;
      else if(format->bigend)
        crossfeed->func = &bs2b_cross_feed_u16be;
      else if (format->sign)
        crossfeed->func = &bs2b_cross_feed_s16le;
      else
        crossfeed->func = &bs2b_cross_feed_u16le;
    }
    else if (format->width == 24) {
      if (format->bigend && format->sign)
        crossfeed->func = &bs2b_cross_feed_s24be;
      else if(format->bigend)
        crossfeed->func = &bs2b_cross_feed_u24be;
      else if (format->sign)
        crossfeed->func = &bs2b_cross_feed_s24le;
      else
        crossfeed->func = &bs2b_cross_feed_u24le;
    }
    else if (format->width == 32) {
      if (format->bigend && format->sign)
        crossfeed->func = &bs2b_cross_feed_s32be;
      else if(format->bigend)
        crossfeed->func = &bs2b_cross_feed_u32be;
      else if (format->sign)
        crossfeed->func = &bs2b_cross_feed_s32le;
      else
        crossfeed->func = &bs2b_cross_feed_u32le;
    }
  }
  else if (format->type == GST_BUFTYPE_FLOAT) {
    if (format->width == 32) {
      if (format->bigend)
        crossfeed->func = &bs2b_cross_feed_fbe;
      else
        crossfeed->func = &bs2b_cross_feed_fle;
    }
    else if (format->width == 64) {
      if (format->bigend)
        crossfeed->func = &bs2b_cross_feed_dbe;
      else
        crossfeed->func = &bs2b_cross_feed_dle;
    }
  }

  if (crossfeed->func == NULL)
    return FALSE;

  GST_CROSSFEED_LOCK (crossfeed);
  gst_crossfeed_update_passthrough (crossfeed);
  GST_CROSSFEED_UNLOCK (crossfeed);

  crossfeed->divider = format->width / 4;

  /* set_rate calls clear, so no need to reset the filter here */
  GST_CROSSFEED_BS2B_LOCK (crossfeed);
  bs2b_set_srate (crossfeed->bs2bdp, format->rate);
  GST_CROSSFEED_BS2B_UNLOCK (crossfeed);

  return TRUE;
}

static void
gst_crossfeed_finalize (GObject * object)
{
  GstCrossfeed *crossfeed = GST_CROSSFEED (object);

  bs2b_close(crossfeed->bs2bdp);
  crossfeed->bs2bdp = NULL;

  g_mutex_free (crossfeed->bs2b_lock);
  g_mutex_free (crossfeed->lock);

  G_OBJECT_CLASS (parent_class)->finalize (object);
}

static gboolean gst_crossfeed_sink_eventfunc (GstBaseTransform * trans,
    GstEvent * event)
{
  GstCrossfeed *crossfeed = GST_CROSSFEED (trans);

  if (GST_EVENT_TYPE (event) == GST_EVENT_NEWSEGMENT) {
    GST_CROSSFEED_BS2B_LOCK (crossfeed);
    bs2b_clear (crossfeed->bs2bdp);
    GST_CROSSFEED_BS2B_UNLOCK (crossfeed);
  }

  return GST_BASE_TRANSFORM_CLASS (parent_class)->event (trans, event);
}

static GstFlowReturn
gst_crossfeed_transform_inplace (GstBaseTransform * base, GstBuffer * outbuf)
{
  GstCrossfeed *crossfeed = GST_CROSSFEED (base);
  void *data = GST_BUFFER_DATA (outbuf);
  gint samples = GST_BUFFER_SIZE (outbuf);

  if(gst_base_transform_is_passthrough (base) ||
      G_UNLIKELY (GST_BUFFER_FLAG_IS_SET (outbuf, GST_BUFFER_FLAG_GAP)))
    return GST_FLOW_OK;

  crossfeed->func (crossfeed->bs2bdp, data, samples / crossfeed->divider);

  return GST_FLOW_OK;
}

static void
gst_crossfeed_update_passthrough (GstCrossfeed * crossfeed)
{
  GstAudioFilter *filter = GST_AUDIO_FILTER (crossfeed);
  GstBaseTransform *trans = GST_BASE_TRANSFORM (crossfeed);
  gboolean passthrough;

  passthrough = (
      filter->format.channels != 2 ||
      !crossfeed->active) ? TRUE : FALSE;

  gst_base_transform_set_passthrough (trans, passthrough);
}

static GType
gst_crossfeed_preset_get_type (void)
{
  static GType crossfeed_preset_type = 0;

  if (!crossfeed_preset_type) {
    static GEnumValue types[] = {
      {
        PRESET_DEFAULT,
        "Closest to virtual speaker placement (30°, 3 meter)   [700Hz, 4.5dB]",
        "default"
      },
      {
        PRESET_CMOY,
        "Close to Chu Moy's crossfeeder (popular)              [700Hz, 6.0dB]",
        "cmoy"
      },
      {
        PRESET_JMEIER,
        "Close to Jan Meier's CORDA amplifiers (little change) [650Hz, 9.0dB]",
        "jmeier"
      },
      {
        PRESET_NONE,
        "No preset",
        "none"
      },
      {0, NULL, NULL},
    };

    crossfeed_preset_type = g_enum_register_static ("GstBs2bPreset", types);
  }

  return crossfeed_preset_type;
}

static void
gst_crossfeed_set_property (GObject * object, guint prop_id,
    const GValue * value, GParamSpec * pspec)
{
  GstCrossfeed *crossfeed = GST_CROSSFEED (object);

  switch (prop_id) {
    case ARG_ACTIVE:
      GST_CROSSFEED_LOCK (crossfeed);
      crossfeed->active = g_value_get_boolean (value);
      gst_crossfeed_update_passthrough (crossfeed);
      /* Clear the filter buffer if it gets set inactive, so we have
       * a fresh start when it gets activated again. */
      if (!crossfeed->active) {
        GST_CROSSFEED_UNLOCK (crossfeed);
        GST_CROSSFEED_BS2B_LOCK (crossfeed);
        bs2b_clear (crossfeed->bs2bdp);
        GST_CROSSFEED_BS2B_UNLOCK (crossfeed);
      } else {
        GST_CROSSFEED_UNLOCK (crossfeed);
      }
      break;
    case ARG_FCUT:
      GST_CROSSFEED_BS2B_LOCK (crossfeed);
      bs2b_set_level_fcut (crossfeed->bs2bdp, g_value_get_int (value));
      GST_CROSSFEED_BS2B_UNLOCK (crossfeed);
      break;
    case ARG_FEED:
      GST_CROSSFEED_BS2B_LOCK (crossfeed);
      bs2b_set_level_feed (crossfeed->bs2bdp, g_value_get_int (value));
      GST_CROSSFEED_BS2B_UNLOCK (crossfeed);
      break;
    case ARG_PRESET:
      switch (g_value_get_enum (value)) {
        case PRESET_DEFAULT:
          GST_CROSSFEED_BS2B_LOCK (crossfeed);
          bs2b_set_level (crossfeed->bs2bdp, BS2B_DEFAULT_CLEVEL);
          GST_CROSSFEED_BS2B_UNLOCK (crossfeed);
          break;
        case PRESET_CMOY:
          GST_CROSSFEED_BS2B_LOCK (crossfeed);
          bs2b_set_level (crossfeed->bs2bdp, BS2B_CMOY_CLEVEL);
          GST_CROSSFEED_BS2B_UNLOCK (crossfeed);
          break;
        case PRESET_JMEIER:
          GST_CROSSFEED_BS2B_LOCK (crossfeed);
          bs2b_set_level (crossfeed->bs2bdp, BS2B_JMEIER_CLEVEL);
          GST_CROSSFEED_BS2B_UNLOCK (crossfeed);
          break;
        default:
          G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
          break;
      }
      break;
    default:
      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
      break;
  }
}

static void
gst_crossfeed_get_property (GObject * object, guint prop_id, GValue * value,
    GParamSpec * pspec)
{
  GstCrossfeed *crossfeed = GST_CROSSFEED (object);

  switch (prop_id) {
    case ARG_ACTIVE:
      g_value_set_boolean (value, crossfeed->active);
      break;
    case ARG_FCUT:
      GST_CROSSFEED_BS2B_LOCK (crossfeed);
      g_value_set_int (value, bs2b_get_level_fcut (crossfeed->bs2bdp));
      GST_CROSSFEED_BS2B_UNLOCK (crossfeed);
      break;
    case ARG_FEED:
      GST_CROSSFEED_BS2B_LOCK (crossfeed);
      g_value_set_int (value, bs2b_get_level_feed (crossfeed->bs2bdp));
      GST_CROSSFEED_BS2B_UNLOCK (crossfeed);
      break;
    case ARG_PRESET:
      GST_CROSSFEED_BS2B_LOCK (crossfeed);
      switch (bs2b_get_level (crossfeed->bs2bdp)) {
        case BS2B_DEFAULT_CLEVEL:
          g_value_set_enum (value, PRESET_DEFAULT);
          break;
        case BS2B_CMOY_CLEVEL:
          g_value_set_enum (value, PRESET_CMOY);
          break;
        case BS2B_JMEIER_CLEVEL:
          g_value_set_enum (value, PRESET_JMEIER);
          break;
        default:
          g_value_set_enum (value, PRESET_NONE);
          break;
      }
      GST_CROSSFEED_BS2B_UNLOCK (crossfeed);
      break;
    default:
      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
      break;
  }
}

static gboolean
plugin_init (GstPlugin * plugin)
{
  return gst_element_register (plugin, "crossfeed", GST_RANK_NONE,
      GST_TYPE_CROSSFEED);
}

GST_PLUGIN_DEFINE (
    GST_VERSION_MAJOR,
    GST_VERSION_MINOR,
    "crossfeed",
    "Improve headphone listening of stereo audio records"
        "using the bs2b library.",
    plugin_init,
    VERSION, "LGPL",
    "gstreamer0.10-bs2b",
    "https://github.com/lazka/gst-bs2b"
);
